deputy.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package deputy provides more advanced options for running commands.
+package deputy
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// ErrorHandling is a flag that tells Deputy how to handle errors running a
+// command.  See the values below for the different modes.
+type ErrorHandling int
+
+const (
+	// DefaultErrs represents the default handling of command errors - this
+	// simply returns the error from Cmd.Run()
+	DefaultErrs ErrorHandling = iota
+
+	// FromStderr tells Deputy to convert the stderr output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStderr
+
+	// FromStdout tells Deputy to convert the stdout output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStdout
+)
+
+// Deputy is a type that runs Commands with advanced options not available from
+// os/exec.  See the comments on field values for details.
+type Deputy struct {
+	// Timeout represents the longest time the command will be allowed to run
+	// before being killed.
+	Timeout time.Duration
+	// Errors describes how errors should be handled.
+	Errors ErrorHandling
+	// StdoutLog takes a function that will receive lines written to stdout from
+	// the command.
+	StdoutLog func(string)
+	// StdoutLog takes a function that will receive lines written to stderr from
+	// the command.
+	StderrLog func(string)
+
+	stderrPipe io.ReadCloser
+	stdoutPipe io.ReadCloser
+}
+
+// Run starts the specified command and waits for it to complete.  Its behavior
+// conforms to the Options passed to it at construction time.
+func (d Deputy) Run(cmd *exec.Cmd) error {
+	if d.StderrLog != nil {
+		var err error
+		d.stderrPipe, err = cmd.StderrPipe()
+		if err != nil {
+			return err
+		}
+	}
+	if d.StdoutLog != nil {
+		var err error
+		d.stdoutPipe, err = cmd.StdoutPipe()
```

Comment:
If both StdoutLog and StderrLog are set, I wonder if it would be
better to just use a single pipe - that way the stdout and stderr
log messages cannot be reordered.